Add group-by feature in APM rules

benakansara opened the pull request:
## Summary
Adds group-by dropdown in the following APM rules.
- APM Latency threshold (Preselected fields: `service.name`, `service.environment`, `transaction.type`)
- APM Failed transaction rate (Preselected fields: `service.name`, `service.environment`, `transaction.type`)
- APM Error count threshold (Preselected fields: `service.name`, `service.environment`)

<img width="609" alt="Screenshot 2023-04-17 at 13 44 34" src="https://user-images.githubusercontent.com/69037875/232475262-41786edf-d16b-4b1f-90a9-8fe242a36bcc.png">

The preselected fields cannot be removed by user. The `transaction.name` field is selectable by user from the group-by dropdown.

- https://github.com/elastic/kibana/issues/154535
- https://github.com/elastic/kibana/issues/154536
- https://github.com/elastic/kibana/issues/154537

Reason message is updated to include group key instead of only service name: 
- https://github.com/elastic/kibana/issues/155011

The `transaction.name` is added to the alert document:
- https://github.com/elastic/kibana/issues/154543

The `transaction.name` action variable is added in UI:
- https://github.com/elastic/kibana/issues/154545

The `transaction.name` is added to the context of active alert notifications:
- https://github.com/elastic/kibana/issues/154547

There is an additional field in group-by dropdown for Error count threshold rule: https://github.com/elastic/kibana/issues/155633
- `error.grouping_key`

## Fixes
- https://github.com/elastic/kibana/issues/154818

### Update on Alert Id
The alert Id is updated for all 3 rules. The new Id is generated from the group key. This is to avoid issues similar to #154818 where alerts are scheduled with same ID. Example of the new alert Ids - `opbeans-java_development_request_GET /flaky`, `opbeans-java_development_GET /fail`

## Out of scope of this PR
- Updating the preview chart based on selected group by fields

## Checklist
- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

## Release note
As the alert Id is updated for the APM Latency threshold rule, APM Failed transaction rate rule and APM Error count rule, the existing alerts, if any, will be recovered, and new alerts will be fired in place of them.

benakansara committed:
add group alerts by feature in apm rules

apmmachine commented:
### :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

Just comment with:
- `/oblt-deploy` : Deploy a Kibana instance using the Observability test environments.
- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)

</p>
</details>

benakansara committed:
fix ci errors

benakansara committed:
fix translations errors

benakansara committed:
fix tests

benakansara committed:
fix tests

benakansara committed:
fix tests

benakansara committed:
fix integration tests

benakansara committed:
update snapshot

benakansara committed:
Merge branch 'main' into feat/add-groupby-in-apm-rules

elasticmachine commented:
Pinging @elastic/apm-ui (Team:APM)

benakansara committed:
Merge branch 'main' into feat/add-groupby-in-apm-rules

benakansara commented:
/oblt-deploy

kdelemme reviewed (commented):
I did a first pass, small neats. Going to tests further

benakansara committed:
add transaction.name in AAD mapping, add transactionName in UI

benakansara commented:
/oblt-deploy

benakansara committed:
update type

benakansara commented on x-pack/plugins/apm/common/rules/schema.ts:
```diff
@@ -15,6 +15,9 @@ export const errorCountParamsSchema = schema.object({
   environment: schema.string(),
+  groupBy: schema.maybe(
+    schema.oneOf([schema.string(), schema.arrayOf(schema.string())])
+  ),
```
done

benakansara commented on x-pack/plugins/apm/public/components/alerting/ui_components/apm_rule_group_by.tsx:
```diff
@@ -0,0 +1,84 @@
+    [onChange, preSelectedOptions]
+  );
+
+  const getPreSelectedOptions = () => {
+    return preSelectedOptions.map((field) => ({
+      label: field,
+      color: 'lightgray',
+      disabled: true,
+    }));
+  };
```
I would rather keep it as function only because it's naming matches with the other function in below where it is called - 

```
const selectedOptions = [
    ...getPreSelectedOptions(),
    ...getUserSelectedOptions(options.groupBy),
  ];
```

benakansara commented on x-pack/plugins/apm/server/routes/alerts/rule_types/error_count/register_error_count_rule_type.ts:
```diff
@@ -86,6 +87,12 @@ export function registerErrorCountRuleType({
       executor: async ({ params: ruleParams, services, spaceId }) => {
+        const predefinedGroupby: string[] = [SERVICE_NAME, SERVICE_ENVIRONMENT];
+
+        const allGroupbyFields: string[] = [
+          ruleParams.groupBy ?? predefinedGroupby,
+        ].flat();
```
We are adding new rule parameter `groupBy`. I added this check to support existing rules when upgrading (without modifying the rule in UI). The existing rules will have `groupBy` undefined so I am adding here predefined group-by fields in that case.
We could always add them, but then we need to use `Set` to filter out duplicates. Because we are always adding default group-by fields in the rule parameters when the rule is saved in UI.
```
new Set([...ruleParams.groupBy, ... predefinedGroupby])
```
Regarding the API, could you tell me how can I test with the API?

benakansara commented:
/oblt-deploy

kpatticha commented:
I used the [scenario](https://github.com/elastic/kibana/blob/main/packages/kbn-apm-synthtrace/src/scenarios/many_transactions_per_service.ts) to generate many unique transaction names to test grouping by `transaction.name`

As we expected it produces many alerts and notifies for the 1000 buckets and in case there are more than 1000 unique transaction names it will alert only for the 1000 (order desc) 


https://user-images.githubusercontent.com/3369346/233348262-40b4bfef-dde7-44f7-95ec-8d50824bbc5d.mov
In the example above I reduced the size to 500 to test it

kpatticha commented on x-pack/plugins/apm/common/rules/schema.ts:
```diff
@@ -15,6 +15,7 @@ export const errorCountParamsSchema = schema.object({
   threshold: schema.number(),
   serviceName: schema.maybe(schema.string()),
   environment: schema.string(),
+  groupBy: schema.maybe(schema.arrayOf(schema.string())),
```
in which cases the `groupBy` can be undefined?

benakansara replied:
For the existing rules, the `groupBy` will be undefined. If Kibana is upgraded with this PR when there are existing APM rules running, `groupBy` will be undefined.

benakansara replied:
yes, the executer checks whether `groupBy` is there. If not, it will add preSelected fields.
```
const allGroupbyFields: string[] = [
  ruleParams.groupBy ?? predefinedGroupby,
].flat();
```

benakansara commented:
> The error I got
> 
> ```
> [2023-04-20T13:19:55.954+02:00][ERROR][plugins.taskManager] Task actions:.slack "44a8548f-df6b-11ed-9646-1b026a973a05" failed: Error: error posting a slack message, retry at 2023-04-20T11:19:56.953Z
> [2023-04-20T13:19:55.964+02:00][WARN ][plugins.actions.slack] action execution failure: .slack:8a3de310-deb2-11ed-9b85-bf06ee56b56c: slack-hooks-alert-notification: error posting a slack message, retry at 2023-04-20T11:19:56.964Z: An HTTP protocol error occurred: statusCode = 429; retry at 2023-04-20T11:19:56.964Z
> ```

The error 429 is for "Too many requests" in short time. I suspect that it is due to number of alerts generated, there are many requests being made to slack.

@simianhacker how can we handle this? Seems to be related to connector.

kpatticha commented on x-pack/plugins/apm/public/components/alerting/ui_components/apm_rule_group_by.tsx:
```diff
@@ -0,0 +1,84 @@
+  errorOptions?: string[];
+}
+
+const fields: string[] = [TRANSACTION_NAME];
```
I expect the fields (options of the combo box) to be defined by the consumer. Similar to `preSelectedOptions`. Currently, all the rules that use the component will have the `transaction.name` option

kpatticha commented on x-pack/plugins/apm/public/components/alerting/ui_components/apm_rule_group_by.tsx:
```diff
@@ -0,0 +1,84 @@
+interface Props {
+  options: { groupBy: string[] | string | undefined };
+  preSelectedOptions: string[];
+  onChange: (groupBy: string | null | string[]) => void;
```
tbh, this is a bit confusing 
`options: { groupBy: string[] | string | undefined };`
`onChange: (groupBy: string | null | string[]) => void;`

why the groupBy can be either a string and string[]?

benakansara replied:
`string` should not be there. I forgot to remove it from here. I will change this.

benakansara replied:
done

kpatticha commented on x-pack/plugins/apm/public/components/alerting/rule_types/transaction_error_rate_rule_type/index.tsx:
```diff
@@ -147,10 +164,44 @@ export function TransactionErrorRateRuleType(props: Props) {
     />
   );
 
+  const groupAlertsBy = (
```
this component looks almost similar to all rules. Can we export it and reuse it?

benakansara replied:
The `helpText` is slightly different for Error count rule as it does not have Transaction Type, that's why I didn't reuse the component.

benakansara committed:
remove string from groupBy

benakansara committed:
Merge branch 'main' into feat/add-groupby-in-apm-rules

kdelemme commented on x-pack/plugins/apm/server/routes/alerts/rule_types/error_count/register_error_count_rule_type.ts:
```diff
@@ -142,32 +144,49 @@ export function registerErrorCountRuleType({
         const errorCountResults =
           response.aggregations?.error_counts.buckets.map((bucket) => {
+            const groupByFields: Record<string, string> = {};
+            bucket.key.forEach((key, keyIndex) => {
+              if (!predefinedGroupby.includes(allGroupbyFields[keyIndex])) {
+                groupByFields[allGroupbyFields[keyIndex]] = key;
+              }
+            });
```
I have some difficulties understanding what this is doing. can you walk me through it please?

benakansara replied:
The idea is that this logic will be useful when we add more fields in the group-by dropdown. Then we don't have to add fields individually. When I implemented this I thought we are going to add many fields in dropdown, not sure now. But I still kept the logic.

```
        services
          .alertWithLifecycle({
            id: bucketKey.join('_'),
            fields: {
              [SERVICE_NAME]: serviceName,
              ...getEnvironmentEsField(environment),
              [TRANSACTION_TYPE]: transactionType,
              [TRANSACTION_NAME]: ruleParams.transactionName,
              [PROCESSOR_EVENT]: ProcessorEvent.transaction,
              [ALERT_EVALUATION_VALUE]: transactionDuration,
              [ALERT_EVALUATION_THRESHOLD]: ruleParams.threshold,
              [ALERT_REASON]: reason,
              ...sourceFields,
              ...groupByFields,
            },
          })
```

kdelemme replied:
I understand now thanks! So basically, groupByFields does not contains the default groupBy keys, since we already added them in the fields object individually. 
But technically since we are doing a spread `...groupByFields`, we could always add them in the groupByFields right?

benakansara replied:
Yes, we could always add them in `groupByFields`. I simply thought not to overwrite the fields. We could then also remove `[SERVICE_NAME]`, `[TRANSACTION_TYPE]` and `...getEnvironmentEsField(environment)` from alert fields, but again I thought not to change existing functionality.
What do you think?

kdelemme replied:
I understand the reluctance to make change in this part of the code: I'll let you decide if you want to change it or not.

IMO, it's easier to always populate the `groupByFields` with all groupBy keys, and then spread it in the `fields` object, than have some non-trivial conditions to exclude some groupBy keys from the `groupByFields`.
But I'm not super knowledgable on this part of the code, and if there is a possibility that the values are different (for example serviceName), it might be better to not change that part :)

benakansara replied:
The environment bucket can be "ENVIRONMENT_NOT_DEFINED" in case this is not present in the source. In this case, we are not adding environment field in the alert doc with this check `...getEnvironmentEsField(environment)`.
I agree we could always add all fields to `groupByFields` and add checks for values that can be different like `service.environment`. But for now, I will keep it as is to avoid any issues with edge cases that we might miss with overwriting fields.

benakansara replied:
@sqren @kdelemme
It looks like the better approach would be to refactor the code for better understandability and to simplify the logic for future.
I have added all group by keys now to `groupByFields` instead of removing existing fields like `serviceName`. I have removed uses of `serviceName`, `environment` and `transactionType`, instead they will come from `groupByFields` itself.
I have also applied necessary checks as it was before, for example,  if environment is undefined, it will not be indexed to alert doc, and the context variable for `environment` will have `Not defined` value as before. The same will apply for any bucket key that is not defined.
I have kept the names of action variables which were previously defined as is - `serviceName`, `environment`, `transactionType` - to support existing rules.

kdelemme reviewed (commented):
Since we changed the groupBy params to be an array, I think we can simplify a bit some conditions here and there.
I also have one question regarding the groupByFields logic in the executors, could you explain me what this is doing please?

benakansara committed:
simplify logic of using groupBy

benakansara commented:
> Since we changed the groupBy params to be an array, I think we can simplify a bit some conditions here and there. I also have one question regarding the groupByFields logic in the executors, could you explain me what this is doing please?

Thanks for catching that. Definitely this could be simplified. I have updated the code.

kdelemme commented on x-pack/plugins/apm/public/components/alerting/rule_types/error_count_rule_type/index.tsx:
```diff
@@ -91,6 +100,13 @@ export function ErrorCountRuleType(props: Props) {
   );
 
+  const onGroupByChange = useCallback(
+    (group: string[] | string | null) => {
```
```suggestion
    (group: string[] | null) => {
```

kdelemme commented on x-pack/plugins/apm/public/components/alerting/rule_types/transaction_duration_rule_type/index.tsx:
```diff
@@ -146,6 +156,13 @@ export function TransactionDurationRuleType(props: Props) {
   );
 
+  const onGroupByChange = useCallback(
+    (group: string[] | string | null) => {
```
```suggestion
    (group: string[] | null) => {
```

kdelemme commented on x-pack/plugins/apm/public/components/alerting/rule_types/transaction_error_rate_rule_type/index.tsx:
```diff
@@ -96,6 +107,13 @@ export function TransactionErrorRateRuleType(props: Props) {
   );
 
+  const onGroupByChange = useCallback(
+    (group: string[] | string | null) => {
```
```suggestion
    (group: string[] | null) => {
```

kdelemme commented on x-pack/plugins/apm/public/components/alerting/ui_components/apm_rule_group_by.tsx:
```diff
@@ -0,0 +1,74 @@
+    }));
+  };
+
+  const getUserSelectedOptions = (groupBy: string[] | undefined) => {
+    return (groupBy ?? [])
+      .filter((group) => !preSelectedOptions.includes(group))
+      .map((field) => ({
+        label: field,
+        color: errorOptions?.includes(field) ? 'danger' : undefined,
+      }));
+  };
```
👍🏻

kdelemme reviewed (commented):
A few type artifacts left (i.e. `string`)

benakansara committed:
remove string type from groupBy

benakansara commented:
> A few type artifacts left (i.e. `string`)

Oh, I thought I got them all. Updated now.

kdelemme reviewed (approved):
Code LGTM.
Tested locally a couple of different scenarios, AAD contains the correct groupByKeys 👍🏻 

<img width="1445" alt="image" src="https://user-images.githubusercontent.com/1376800/233457106-3a0cc74b-3fb0-4553-bae5-1360a8e1cf2a.png">
<img width="1449" alt="image" src="https://user-images.githubusercontent.com/1376800/233457195-449b4dcf-99fd-419b-9847-30d96a1ca81a.png">

benakansara commented:
/oblt-deploy

simianhacker commented:
> The error 429 is for "Too many requests" in short time. I suspect that it is due to number of alerts generated, there are many requests being made to slack.
>
> @simianhacker how can we handle this? Seems to be related to connector.

I assume there is a rate limit on Slack messages.  For a group by combination that has the potential to trigger a high volume of actions, we should recommend to the user to change the "action frequency" to "Summary of alerts".  Maybe in a follow up PR we could add a warning message if the combined cardinality of each field is above 100 (or whatever ceiling makes sense).

@elastic/response-ops Thoughts?

kpatticha commented:
/oblt-deploy

kpatticha commented on x-pack/plugins/apm/public/components/alerting/rule_types/error_count_rule_type/index.tsx:
```diff
@@ -138,11 +154,44 @@ export function ErrorCountRuleType(props: Props) {
     />
   );
 
+  // const fields: string[] = [TRANSACTION_NAME];
```
oops 👀

benakansara replied:
removed commented code

kpatticha commented on x-pack/plugins/apm/public/components/alerting/ui_components/apm_rule_group_by.tsx:
```diff
@@ -0,0 +1,74 @@
+        defaultMessage: 'Everything',
+      })}
+      aria-label={i18n.translate('xpack.apm.ruleFlyout.groupByAriaLabel', {
+        defaultMessage: 'Graph per',
```
```suggestion
        defaultMessage: 'Group by',
```

benakansara committed:
remove commented code

benakansara committed:
Update x-pack/plugins/apm/public/components/alerting/ui_components/apm_rule_group_by.tsx

Co-authored-by: Katerina Patticha <kate@kpatticha.com>

mikecote commented:
> I assume there is a rate limit on Slack messages. For a group by combination that has the potential to trigger a high volume of actions, we should recommend to the user to change the "action frequency" to "Summary of alerts". Maybe in a follow up PR we could add a warning message if the combined cardinality of each field is above 100 (or whatever ceiling makes sense).
> 
> https://github.com/orgs/elastic/teams/response-ops Thoughts?

The assumption of rate limiting is correct, Slack will even tell us when to try again and the actions framework will retry at the given time. Seems fair to recommend alert summaries whenever the cardinality may be high, or even provide it as a default at times. @XavierM do you have some thoughts for the UX?

kpatticha reviewed (approved):
LGTM great work 🎉 
Thank you also for fixing the alert id

benakansara committed:
Merge branch 'main' into feat/add-groupby-in-apm-rules

benakansara commented:
/oblt-deploy

benakansara commented on x-pack/plugins/apm/server/routes/alerts/rule_types/error_count/register_error_count_rule_type.ts:
```diff
@@ -145,37 +146,49 @@ export function registerErrorCountRuleType({
               windowSize: ruleParams.windowSize,
               windowUnit: ruleParams.windowUnit,
+              group: bucketKey
+                .filter((key) => !key.includes('NOT_DEFINED'))
```
yes, it is for any field that is not defined. The group here is used to create reason message. If we leave values like `ENVIRONMENT_NOT_DEFINED_VALUE`, the reason message will be for example - `Avg. latency is 1,032 ms in the last 5 mins for opbeans-node, ENVIRONMENT_NOT_DEFINED, request. Alert when > 15 ms.`. Isn't it better with only `opbeans-node, request`?

sqren replied:
Firstly, I think we should pass the `group` unchanged and then handle this `formatErrorCountReason`. Secondly, by removing it entirely the user won't be able to see which group this alert is for. The alert is indeed for the environment `undefined`. 

Secondly, if we want to format `ENVIRONMENT_NOT_DEFINED_VALUE` we can do something like:

```ts
if (value === ENVIRONMENT_NOT_DEFINED_VALUE) {
  return i18n.translate('xpack.apm.filter.environment.notDefinedLabel', {
    defaultMessage: 'Not defined',
  });
}
```

benakansara replied:
We can format the undefined values. My concern was that showing undefined environment in reason message may not be useful, because the environment is not defined in the source, so user doesn't use that field.

I will change the code to include undefined values in reason message.

benakansara replied:
Updated reason message:
`Avg. latency is 1,032 ms in the last 5 mins for opbeans-node, environment not defined, request. Alert when > 15 ms.`

benakansara commented on x-pack/plugins/apm/server/routes/alerts/rule_types/get_group_action_variables.ts:
```diff
@@ -0,0 +1,21 @@
+): Record<string, string> => {
+  return Object.keys(groupByFields).reduce<Record<string, string>>(
+    (acc, cur) => {
+      acc[cur === TRANSACTION_NAME ? 'transactionName' : cur] =
```
This is for matching the naming of action variable `transactionName` that was already defined.

sqren replied:
> This is for matching the naming of action variable transactionName that was already defined.

Anything stopping us from changing it. It was introduced in this release cycle so it has not yet been released (won't be a breaking change)

benakansara replied:
It would be easier to use `transaction.name` (without extra step of renaming it). Looking at the other variable names, such as serviceName, transasctionType, at the time,  it also seemed to be consistent to name variables like this.

I will change the action variable to be `transaction.name`.

benakansara replied:
done

benakansara committed:
Update x-pack/plugins/apm/server/routes/alerts/rule_types/error_count/register_error_count_rule_type.ts

Co-authored-by: Søren Louv-Jansen <sorenlouv@gmail.com>

benakansara committed:
remove explicit type

benakansara committed:
code review changes, refactoring

benakansara commented on x-pack/plugins/apm/server/routes/alerts/rule_types/error_count/register_error_count_rule_type.ts:
```diff
@@ -145,37 +146,49 @@ export function registerErrorCountRuleType({
+
             const [serviceName, environment] = bucket.key;
+
+            const bucketKey = bucket.key;
```
Refactored the code which solves this too.

benakansara committed:
renaming transactionName

benakansara committed:
updated comment

benakansara committed:
fix tests

sqren commented on x-pack/plugins/apm/common/rules/apm_rule_types.ts:
```diff
@@ -40,22 +40,36 @@ const THRESHOLD_MET_GROUP: ActionGroup<ThresholdMetActionGroupId> = {
+const formatGroupKey = (groupKey: string[]) =>
+  ` for ${groupKey
+    .map((key) =>
+      key.includes('NOT_DEFINED') ? getValueForNotDefinedField(key) : key
```
We already have some helpers in `x-pack/plugins/apm/common/environment_filter_values.ts`.

```suggestion
      key === ENVIRONMENT_NOT_DEFINED_VALUE ? getEnvironmentLabel(key) : key
```

Btw. we should avoid doing `key.includes('NOT_DEFINED')` because it makes it harder to reason about. If we want custom formatting we should do it for specific values (eg `ENVIRONMENT_NOT_DEFINED_VALUE`)

benakansara replied:
This is to address all group keys that are not defined. I think it is better to generalize it than checking each key individually. If we have more group keys in future, then checking `NOT_DEFINED` is more readable, IMO.
I didn't use `environment_filter_values.ts` because the naming was specific to `environment`, whereas I have applied this to all group keys.

sqren commented on x-pack/plugins/apm/server/routes/alerts/rule_types/error_count/register_error_count_rule_type.test.ts:
```diff
@@ -142,25 +241,225 @@ describe('Error count alert', () => {
       interval: '5 mins',
       viewInAppUrl:
         'http://localhost:5601/eyr/app/apm/services/foo/errors?environment=env-foo',
+      'transaction.name': 'Not defined',
```
This looks wrong. If `transaction.name` is not specified as grouping param it should just be `undefined` (not be there)

benakansara replied:
I followed the existing pattern where if environment is not specified, the `environment` context variable had `Not defined` value. The same behaviour is now applied to all group keys. If we do not include undefined values in context variables, it will behave differently than earlier versions.

benakansara replied:
Ah, I see what you mean. Actually, in this case, the `transaction.name` is specified in the grouping parameter, but it is not defined in the source. The bucket key will contain `TRANSACTION_NAME_NOT_DEFINED`, similar to when `environment` is not defined.

If `transaction.name` is not specified in the grouping param, then we are not adding it anywhere.

sqren commented on x-pack/plugins/apm/server/routes/alerts/rule_types/error_count/register_error_count_rule_type.test.ts:
```diff
@@ -126,11 +130,106 @@ describe('Error count alert', () => {
-      'apm.error_rate_foo_env-foo',
-      'apm.error_rate_foo_env-foo-2',
-      'apm.error_rate_bar_env-bar',
+      'foo_env-foo_TRANSACTION_NAME_NOT_DEFINED',
```
This should just be:

```suggestion
      'foo_env-foo,
```

But perhaps this point to another issue: perhaps what we actually want to pass in is:
```json
{
  `service.name`: `foo`,
  `service.environment`: `foo_env`
}
```

This way we know which fields the values belong to. And we can format them accordingly (eg formatting environment with `getEnvironmentLabel` from x-pack/plugins/apm/common/rules/apm_rule_types.ts)

benakansara replied:
This is only for alert ID (not used anywhere user facing). Not sure why we should format it here?

sqren replied:
Either way, `TRANSACTION_NAME_NOT_DEFINED` should not be a thing.

benakansara replied:
I formatted the reason message as per your suggestion. Previously I was not adding undefined fields in the reason message.

I saw `TRANSACTION_NAME_NOT_DEFINED` for missing transaction name bucket on CCS.

sqren commented on x-pack/plugins/apm/server/routes/alerts/rule_types/utils/get_groupby_terms.ts:
```diff
@@ -0,0 +1,21 @@
+import { ENVIRONMENT_NOT_DEFINED } from '../../../../../common/environment_filter_values';
+import { SERVICE_ENVIRONMENT } from '../../../../../common/es_fields/apm';
+
+export const getGroupByTerms = (groupBy: string[] | undefined) => {
+  return (groupBy ?? []).map((group) => {
+    return {
+      field: group,
+      missing:
+        group === SERVICE_ENVIRONMENT
+          ? ENVIRONMENT_NOT_DEFINED.value
+          : group.replaceAll('.', '_').toUpperCase().concat('_NOT_DEFINED'),
+    };
+  });
+};
```
The "Not defined" behaviour is specifically for `service.environment`. It is technical debt and ideally we want to remove it in the future. We do not want to apply this behaviour to other fields.

```suggestion
export const getGroupByTerms = (groupByFields: string[] | undefined = []) => {
  return groupByFields.map((groupByField) => {
    return {
      field: groupByField,
      missing:
        groupByField === SERVICE_ENVIRONMENT
          ? ENVIRONMENT_NOT_DEFINED.value
          : undefined,
    };
  });
};
```

Btw I updated the naming to align with the caller (`groupByFields` instead of `groupBy`) abd added a default param so we don't have to handle that in the function body.

benakansara replied:
If we remove `missing: ENVIRONMENT_NOT_DEFINED` from the `multi_terms` aggregation, and if the `environment` field is not there, will there be still a bucket key for `environment`? This will need further testing.

I wasn't aware this was a technical debt. I think we should address all technical debt related to fields being not defined in a separate PR. What do you think?

sqren replied:
> If we remove `missing: ENVIRONMENT_NOT_DEFINED` from the multi_terms aggregation

We should not remove `missing: ENVIRONMENT_NOT_DEFINED` when field is `SERVICE_ENVIRONMENT`. On the contrary: we should **only** have a missing bucket for environment.

benakansara replied:
I saw the missing bucket for transaction name too when testing on CCS. What is the disadvantage of keeping NOT_DEFINED for every field in aggregation?

sqren commented:
@benakansara We only want to have a `missing` bucket for `service.environment`:

### Example 1

- Group-By: [`service.name`, `service.environment`, `transaction.type`]
- aggegation: [
      { field: 'service.name' },
      { field: 'service.environment', missing: 'ENVIRONMENT_NOT_DEFINED' },
      { field: 'transaction.type' },
    ]
- Reason: `Avg. latency is 1,032 ms in the last 5 mins for opbeans-node, environment not defined, request. Alert when > 15 ms.`
- Alert ID: `opbeans-node_ENVIRONMENT_NOT_DEFINED_request`
- Context: `environment`: `Not defined`
- Alert document (AAD): The `environment` field is NOT indexed

### Example 2

- Group-By: [`service.name`, `service.environment`, `transaction.name`]
- aggegation: [
      { field: 'service.name' },
      { field: 'service.environment', missing: 'ENVIRONMENT_NOT_DEFINED' },
      { field: 'transaction.name' },
    ]
- Reason: `Error count is 5 in the last 5 mins for opbeans-node, development. Alert when > 2.`
- Alert ID: `opbeans-node_development`
- Context: `transaction.name`: `Not defined`
- Alert document (AAD): The `transaction.name` field is NOT indexed

benakansara commented:
@sqren and I discussed offline, and we agreed on the below expected behaviour.

### Example 1

- Group-By: [`service.name`, `service.environment`, `transaction.type`]
- aggegation: [
      { field: 'service.name' },
      { field: 'service.environment', missing: 'ENVIRONMENT_NOT_DEFINED' },
      { field: 'transaction.type' },
    ]
- Reason: `Avg. latency is 1,032 ms in the last 5 mins for service: opbeans-node, env: Not defined, type: request. Alert when > 15 ms.`
- Alert ID: `opbeans-node_ENVIRONMENT_NOT_DEFINED_request`
- Context: `environment`: `Not defined`
- Alert document (AAD): The `environment` field is NOT indexed

### Example 2

- Group-By: [`service.name`, `service.environment`, `transaction.name`]
- aggegation: [
      { field: 'service.name' },
      { field: 'service.environment', missing: 'ENVIRONMENT_NOT_DEFINED' },
      { field: 'transaction.name' },
    ]
- Reason: `Error count is 5 in the last 5 mins for service: opbeans-node, env: development, name: foo. Alert when > 2.`
- Alert ID: `opbeans-node_development_foo`
- Context: `transaction.name`: `foo`
- Alert document (AAD): `transaction.name`: `foo`

### Example 3

- Group-By: [`service.name`, `service.environment`, `transaction.name`]
- aggegation: [
      { field: 'service.name' },
      { field: 'service.environment', missing: 'ENVIRONMENT_NOT_DEFINED' },
      { field: 'transaction.name' },
    ]
- If there is no `transaction.name` in the document, there will not be an alert in this case.

benakansara committed:
code review changes

benakansara committed:
Merge branch 'main' into feat/add-groupby-in-apm-rules

benakansara commented:
I can add in this PR.
Could you confirm what should be the reason message format with new fields?
<img width="738" alt="Screenshot 2023-04-24 at 15 40 24" src="https://user-images.githubusercontent.com/69037875/234014196-7074570d-ac0a-4381-9c50-f8daa753ef12.png">

benakansara committed:
add two new group by fields in error count rule

benakansara committed:
add error grouping fields to index mapping

benakansara committed:
Merge branch 'main' into feat/add-groupby-in-apm-rules

benakansara commented on x-pack/plugins/apm/server/routes/alerts/register_apm_rule_types.ts:
```diff
@@ -50,6 +52,14 @@ export const apmRuleTypeAlertFieldMap = {
+  [ERROR_GROUP_NAME]: {
+    type: 'text',
+    required: false,
+  },
```
@sqren I would like to get your opinion on adding these fields in the index mapping and whether their types are accurate.

benakansara committed:
change error.grouping_name type to keyword

benakansara committed:
update snapshot

benakansara committed:
Merge branch 'main' into feat/add-groupby-in-apm-rules

sqren commented on x-pack/plugins/apm/common/rules/apm_rule_types.ts:
```diff
@@ -40,95 +49,134 @@ const THRESHOLD_MET_GROUP: ActionGroup<ThresholdMetActionGroupId> = {
+        groupByFields[field]
+      )}`
+  );
+  return ` for ${groupByFieldLabels.join(', ')}`;
```
Shouldn't "for" be translated and thus defined as part of the string?

>     defaultMessage: `Error count is {measured} in the last {interval} for {group}. Alert when > {threshold}.`,



```suggestion
  return groupByFieldLabels.join(', ');
```

sqren commented on x-pack/plugins/apm/common/rules/apm_rule_types.ts:
```diff
@@ -40,95 +49,134 @@ const THRESHOLD_MET_GROUP: ActionGroup<ThresholdMetActionGroupId> = {
 }) {
   return i18n.translate('xpack.apm.alertTypes.errorCount.reason', {
-    defaultMessage: `Error count is {measured} in the last {interval} for {serviceName}. Alert when > {threshold}.`,
+    defaultMessage: `Error count is {measured} in the last {interval}{group}. Alert when > {threshold}.`,
```
As mentioned above
```suggestion
    defaultMessage: `Error count is {measured} in the last {interval} for {group}. Alert when > {threshold}.`,
```

benakansara replied:
done

benakansara committed:
Update x-pack/plugins/apm/common/rules/apm_rule_types.ts

Co-authored-by: Søren Louv-Jansen <sorenlouv@gmail.com>

benakansara committed:
update reason message for translation

sqren commented on x-pack/plugins/apm/server/routes/alerts/action_variables.ts:
```diff
@@ -71,7 +71,7 @@ export const apmActionVariables = {
       { defaultMessage: 'The transaction name the alert is created for' }
     ),
-    name: 'transactionName' as const,
+    name: 'transaction.name' as const,
```
I suggested using the ECS field names, like you did here, but I'm getting second thoughts now that I see we've camelcased the other params (`transactionType` and `serviceName`).

@dgieselaar any thoughts?

dgieselaar replied:
if anything we should keep it consistent. I'm honestly not sure whether dots in field names are well supported in the templating engine used by the alerting framework.

benakansara replied:
It is possible to use dot separated fields in the action template.

<img width="575" alt="Screenshot 2023-04-25 at 12 04 38" src="https://user-images.githubusercontent.com/69037875/234244439-6c040729-bb9b-4dc4-8fc4-1b239bf8ee38.png">

dgieselaar replied:
But aren't we setting `context['error.grouping_name']` rather than `context.error.grouping_name`?

benakansara replied:
yes, we are setting `error.grouping_name` in the `context`. So the way to access it in template is using `context.error.grouping_name`.

dgieselaar replied:
Is the alerting framework flattening fields? E.g., there's no difference between `{ 'error.grouping_name': .... }` and `{ error: { grouping_name: ... } }`?

dgieselaar replied:
Hmm that's not great, I would rather not depend on it here because it feels a bit hackish, but as long as we have tests to cover it (ie, the template gets rendered with the right variables) I don't mind.

benakansara replied:
I don't mind changing the names. We should decide on naming now because we cannot modify it in next version or it will break backward compatibility.

@sqren should I keep `transaction.name` or change to `transactionName`? (same for error.grouing_key, error.grouping_name)

sqren replied:
Yes, let's change back to camel cased (`transactionName`). Sorry for the noise

sqren commented on x-pack/test/apm_api_integration/tests/alerts/transaction_error_rate.spec.ts:
```diff
@@ -0,0 +1,169 @@
+      });
+
+      it('checks if alert is active', async () => {
+        const executionStatus = await waitForRuleStatus({
```
Like mentioned above: can you also verify that the group-by params are added to the alert as expected?

benakansara replied:
done

benakansara commented:
> Is this a valid id? Spaces and quotes are normally not allowed in ids. We may have to hash the message. @dgieselaar you had some thoughts/concerns about id generation.

I was not sure about that either, so wanted to get your feedback on it.

sqren commented:
> Should I hide `error.grouping_name` from the UI (but keeping the logic, tests, etc.) until we decide on the alert Id format?

Hmm.. Yeah, you could do that. But alternatively you could hash the message like:

```ts
export function getHash(content: string) {
  return createHash('sha256').update(content).digest('hex');
}
```

benakansara committed:
Update x-pack/plugins/apm/server/routes/alerts/rule_types/error_count/register_error_count_rule_type.ts

Co-authored-by: Søren Louv-Jansen <sorenlouv@gmail.com>

sqren commented on x-pack/plugins/apm/server/routes/alerts/rule_types/utils/get_groupby_alert_fields.test.ts:
```diff
@@ -0,0 +1,39 @@
+import { getAlertFieldsFromGroupBy } from './get_groupby_alert_fields';
+
+describe('getAlertFieldsFromGroupBy', () => {
+  it('should not contain service.environment', () => {
+    const result = getAlertFieldsFromGroupBy({
+      'service.name': 'opbeans-java',
+      'service.environment': 'ENVIRONMENT_NOT_DEFINED',
+      'transaction.name': 'tx-java',
+    });
+    expect(result).toMatchInlineSnapshot(`
+      Object {
+        "service.name": "opbeans-java",
+        "transaction.name": "tx-java",
+      }
+    `);
```
> I can change that. I will index `service.environment`: `Not defined` in this case (same as context variable). Let me know if the text should be changed.

We should index it as `service.environment`: `ENVIRONMENT_NOT_DEFINED`

sqren replied:
> I can change that. I will index `service.environment`: `Not defined` in this case (same as context variable). Let me know if the text should be changed.

We should index it as `service.environment`: `ENVIRONMENT_NOT_DEFINED` (not the formatted string)

benakansara replied:
done

benakansara committed:
update tests

benakansara commented on x-pack/test/apm_api_integration/tests/alerts/error_count_threshold.spec.ts:
```diff
@@ -93,13 +96,19 @@ export default function ApiTest({ getService }: FtrProviderContext) {
             threshold: 1,
             windowSize: 1,
             windowUnit: 'h',
+            groupBy: ['service.name', 'service.environment', 'transaction.name'],
```
done

benakansara committed:
index environment in alert doc also when not defined

benakansara commented:
> Hmm.. Yeah, you could do that. But alternatively you could hash the message like:

I am fine with either way (removing for now or using hash). My only concern is that if we need more discussion around alert ID, we should wait till it is clear.

dgieselaar commented on x-pack/plugins/apm/server/routes/alerts/rule_types/error_count/register_error_count_rule_type.ts:
```diff
@@ -149,40 +146,42 @@ export function registerErrorCountRuleType({
-            const [serviceName, environment] = bucket.key;
+            const groupByFields = bucket.key.reduce(
+              (obj, bucketKey, bucketIndex) => {
+                obj[allGroupbyFields[bucketIndex]] = bucketKey;
```
I'm wary of using the index here. Are we enforcing the order somewhere?

benakansara replied:
The `bucket.key` contains ONLY the values in the order the group by fields are added in the rule. It doesn't have the field names (like `service.name`: `opbeans-go`), so we need to reply on the index to generate this mapping.

dgieselaar replied:
The multi_terms agg doesn't return it, but you're telling the multi_terms agg the terms (and the order), so you should be able to map the field names to the bucket key.

benakansara replied:
Yes, that is what I am doing here with `allGroupbyFields` which is also used in mulit_terms agg.

benakansara committed:
update snapshot

dgieselaar commented on x-pack/plugins/apm/server/routes/alerts/rule_types/error_count/register_error_count_rule_type.ts:
```diff
@@ -88,6 +88,9 @@ export function registerErrorCountRuleType({
       isExportable: true,
       executor: async ({ params: ruleParams, services, spaceId }) => {
+        const predefinedGroupby = [SERVICE_NAME, SERVICE_ENVIRONMENT];
+        const allGroupbyFields = ruleParams.groupBy ?? predefinedGroupby;
```
Users can (and do) use the API directly though. I think these kinds of things should ideally be in the server. Part of it in the browser and part of it in the server is a bit risky IMHO. But not a blocker.

benakansara committed:
always adding predefined group bys

benakansara commented:
@dgieselaar @sqren 
I am going to remove `error.grouping_name` entirely (UI, logic) for now. I will add it back when alert ID part is clear.

benakansara committed:
remove error.grouping_name groupby in error count rule

sqren commented on x-pack/test/apm_api_integration/tests/alerts/transaction_error_rate.spec.ts:
```diff
@@ -0,0 +1,186 @@
+    let ruleId: string;
+    let actionId: string | undefined;
+
+    const INDEX_NAME = 'transaction-error-rate';
```
I saw we are using the same variable name in other tests. Can you change it to be more descriptive? If it is the index where the index action documents are stored we can call it `ALERT_ACTION_INDEX_NAME`

```suggestion
    const INDEX_NAME = 'transaction-error-rate';
```

sqren commented on x-pack/test/apm_api_integration/tests/alerts/transaction_duration.spec.ts:
```diff
@@ -0,0 +1,182 @@
+      await supertest.delete(`/api/actions/connector/${actionId}`).set('kbn-xsrf', 'foo');
+      await esDeleteAllIndices([INDEX_NAME]);
+      await es.deleteByQuery({
+        index: '.alerts*',
```
Perhaps extract it and make it more specific:

```ts
const APM_ALERTS_INDEX = '.alerts-observability.apm.alerts-default';
```

```suggestion
        index: APM_ALERTS_INDEX,
```

benakansara replied:
done

benakansara committed:
rename back action variables

sorenlouv reviewed (approved):
Just a few nits left. LGTM and thanks for all the help!
APM alerting is improving rapidly these days!

benakansara committed:
updated tests

benakansara committed:
updated tests

sqren commented on x-pack/test/apm_api_integration/tests/alerts/error_count_threshold.spec.ts:
```diff
@@ -170,6 +170,10 @@ export default function ApiTest({ getService }: FtrProviderContext) {
         expect(resp.hits.hits[0]._source).property('transaction.name', 'tx-java');
+        expect(resp.hits.hits[0]._source).property(
+          'error.grouping_key',
+          '2cb494454edc437679309dc190d692368ae7da490f45aff291fb6f7989437d4a'
```
You can avoid hardcoding the error grouping key like:

```
const errorMessage = '[ResponseError] index_not_found_exception';
const errorGroupingKey = getErrorGroupingKey(errorMessage);

```suggestion
          errorGroupingKey
```

benakansara replied:
thanks! updated.

benakansara committed:
remove hardcoded error grouping key

sqren commented on x-pack/plugins/apm/server/routes/alerts/rule_types/transaction_duration/register_transaction_duration_rule_type.ts:
```diff
@@ -105,6 +103,16 @@ export function registerTransactionDurationRuleType({
+        TRANSACTION_TYPE,
+      ];
+
+      const allGroupbyFields = Array.from(
+        new Set([...predefinedGroupby, ...(ruleParams.groupBy ?? [])])
+      );
```
nit: this can be simplified like:

```ts
      const allGroupbyFields = union(predefinedGroupBy, ruleParams.groupBy)
```

(also: there is a consistency with "group by". Sometimes it's written as "groupby" other times as "groupBy". I suggest we consistently use the latter.

benakansara replied:
I will address this in a new PR.
